### PR TITLE
Fixes #4372

### LINKED ISSUE #4372: chai definitions are lacking `should`
I have an [update for the chai definitions] <sup>†</sup> that add `should` support (both the [`should` fluent] method and the [`should` extras]).

The `chai-tests.js` file has ~460 `expect` tests. Do I need to replicate each of those as a `should` test in order to submit the PR?

<sup>†</sup> on branch [AGBrown//DefinitelyTyped-ABContrib/4372-chai-should]

[update for the chai definitions]: https://github.com/AGBrown/DefinitelyTyped-ABContrib/commit/2e1426a "AGBrown//DefinitelyTyped-ABContrib/2e1426a"
[AGBrown//DefinitelyTyped-ABContrib/4372-chai-should]: https://github.com/borisyankov/DefinitelyTyped/compare/master...AGBrown:4372-chai-should "PR compare for 4372-chai-should"
[`should` fluent]: http://chaijs.com/guide/styles/#should "chaijs > guide > styles > should"
[`should` extras]: http://chaijs.com/guide/styles/#should-extras "chaijs > guide > styles > should extras"

### PR DESCRIPTION
**Pull request by AGBrown**
PR for #4372

### CODE REVIEW

**Comment by AGBrown** [comment]
@jedmao, @vvakame I fixed the final issue in this PR - it was actually due to a conflicting `should` definition in a chai plugin file which I've now removed ([chai-subset.d.ts](https://github.com/AGBrown/DefinitelyTyped-ABContrib/commit/8fa7bd02f9db65c6d39ec82debed6c03297b8d3a#diff-2faf01fab0fa8d83fffe812879736a03)).

File: chai/chai-tests.ts [hunk 1]
```diff
@@ -5,290 +5,416 @@ import chai = require('chai');
...
     expect(new Object()).to.be.a('object');
+    (new Object()).should.be.a('object');
     expect({}).to.be.a('object');
+    ({}).should.be.a('object');
     expect([]).to.be.a('array');
+    [].should.be.a('array');
     expect(() => { }).to.be.a('function');
+    (() => { }).should.be.a('function');
     expect(null).to.be.a('null');
+    // has no should equivalent?
```
Comment (jedmao) [inline chai/chai-tests.ts]: That would make sense. See MDN for the [difference between null and undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/null#Difference_between_null_and_undefined):
```js
typeof null        // object (bug in ECMAScript, should be null)
typeof undefined   // undefined
null === undefined // false
null  == undefined // true
```
Also, you can run this in your console:
```js
(null).prototype // Uncaught: TypeError: Cannot read property 'prototype' of null
```

File: chai/chai.d.ts [hunk 2]
```diff
@@ -25,6 +28,28 @@ declare module Chai {
         (target: any, message?: string): Assertion;
     }
 
+    interface ShouldAssertion {
+        equal(val1: any, val2: any, msg?: string): void;
```
Comment (jedmao) [inline chai/chai.d.ts]: I think `value1`, `value2` and `message` would be more consistent with the rest of the definition file.

File: chai/chai.d.ts [hunk 3]
```diff
@@ -25,6 +28,28 @@ declare module Chai {
...
+        Throw: ShouldThrow;
+        throw: ShouldThrow;
```
Comment (jedmao) [inline chai/chai.d.ts]: Where is the lowercase `throw` mentioned in the chai docs? I can't find it. Did you find this in the source code or something?

File: chai/chai.d.ts [hunk 4]
```diff
@@ -25,6 +28,28 @@ declare module Chai {
...
+        exist(val: any, msg?: string): void;
```
Comment (jedmao) [inline chai/chai.d.ts]: `value` and `message` here, as mentioned above.

File: chai/chai.d.ts [hunk 5]
```diff
@@ -25,6 +28,28 @@ declare module Chai {
...
+    }
+
+    interface Should extends ShouldAssertion {
+        not: ShouldAssertion;
+        fail(actual: any, expected: any, message?: string, operator?: string): void;
+    }
+
+    interface ShouldThrow {
+        (actual: Function): void;
+        (actual: Function, expected: string, message?: string): void;
```
Comment (jedmao) [inline chai/chai.d.ts]: You can use a union type here to remove the line below it by defining `expected: string|RegExp`. I'm sure there are other opportunities for this in this file as well, including code you didn't write. This is a fairly fresh feature, so it doesn't surprise me that it doesn't currently show in this file, but it's definitely something we should be doing from here on out.

File: chai/chai.d.ts [hunk 6]
```diff
@@ -25,6 +28,28 @@ declare module Chai {
...
+        (actual: Function, expected: RegExp, message?: string): void;
+        (actual: Function, constructor: Error, expected?: string, message?: string): void;
```
Comment (jedmao) [inline chai/chai.d.ts]: Same here. We can do `constructor: Error|Function, expected?: string|RegExp`

File: chai/chai.d.ts [hunk 7]
```diff
@@ -281,3 +306,7 @@ declare var chai: Chai.ChaiStatic;
 declare module "chai" {
     export = chai;
 }
+
+interface Object {
+    should: Chai.Assertion;
```
Comment (jedmao) [inline chai/chai.d.ts]: Thanks for figuring this out! I'm so glad you found and removed the conflict!

File: chai/chai.d.ts (hunk 3 above)
Comment (AGBrown) [inline chai/chai.d.ts]: I think they are here: https://github.com/chaijs/chai/blob/2.3.0/lib/chai/interface/should.js#L90-L91
```javascript
    should['throw'] = should['Throw'];
    should.not['throw'] = should.not['Throw'];
```

**Comment by jedmao** [comment]
@AGBrown, code review is complete, Sir. Thanks so much for doing this! As soon as you address the super-small issues I've stated above I'll give you a :+1: .

**Comment by jedmao** [comment]
Oh, and just push another commit before you squash it down into one so the changes are easier to review. Thanks!

File: chai/chai.d.ts (hunk 2 above)
Comment (AGBrown) [inline chai/chai.d.ts]: I can change that (I was just going straight from the [source code](https://github.com/chaijs/chai/blob/2.3.0/lib/chai/interface/should.js#L63), didn't stop to think if it was well worded)

File: chai/chai.d.ts (hunk 4 above)
Comment (AGBrown) [inline chai/chai.d.ts]: Happy with that - will make the change ([source code](https://github.com/chaijs/chai/blob/2.3.0/lib/chai/interface/should.js#L71) again)

**Comment by jedmao** [comment]
BTW – you can edit the original message/description for this PR to say "Fixes #4372" and it will automatically close the associated issue when this is merged.

File: chai/chai.d.ts (hunk 5 above)
Comment (AGBrown) [inline chai/chai.d.ts]: Happy with that - hadn't done it to stick with the current file conventions, but will make that change (and for `constructor: Error|Function` in the same set of overloads)

File: chai/chai.d.ts (hunk 6 above)
Comment (AGBrown) [inline chai/chai.d.ts]: Ah, yes, snap - see reply to previous comment.

File: chai/chai-tests.ts (hunk 1 above)
Comment (AGBrown) [inline chai/chai-tests.ts]: That was a badly written rhetorical comment. I have reworded.

File: chai/chai.d.ts (hunk 7 above)
Comment (AGBrown) [inline chai/chai.d.ts]: mcve, friends, mcve :smile:

**Comment by AGBrown** [comment]
Ignore c7f4d99 - something bad happened to the line endings

File: chai/chai.d.ts (hunk 4 above)
Comment (jedmao) [inline chai/chai.d.ts]: Yeah. I see arg names like this all the time in libraries, but I don't agree with it. Especially when we're dealing with TypeScript, we're trying to get more information out of libraries, not less. These short-hand names can be confusing at times; though, not particularly in this case. I think the longer name serves well in Intellisense.

**Comment by jedmao** [comment]
Looks good! Build passes. Go ahead and squash everything into 1 commit please.

**Comment by AGBrown** [comment]
@jedmao thanks for the code review; squashed and pushed.

**Comment by jedmao** [comment]
:+1:

**Comment by vvakame** [comment]
@jedmao thanks!
@AGBrown thanks mate!

**Pull request merged**